DESIGN.md

```diff
@@ -28,6 +28,196 @@ daemon is spawned to handle future requests in its stead.
+or a null version.
+
+The metadata backend is responsible for handling the create, overwrite and
+delete operations of version metadata upon receiving instructions. We use
+version identifiers as the suffix for the keys of the object versions, and a
+special version to represent the latest version.
+
+The following is a summary of the design and details of implementation of
+versioning in S3.
+
+#### Version Id Format
+
+The version id is generated by the metadata backend, and encoded in a
+hexadecimal string format by S3. The following information is encoded in a
+version ID:
+
+`object_name separator version_id`
+where:
+
+- `object_name`: is the key of the object in metadata
+- `separator`: we use the `null` character (`0x00` or `\0`) as the separator
+  between the `object_name` and the `version_id` of a key
+- `version_id`: is the version identifier; this encodes the ordering
+  information in the texture format as metadata orders keys alphabetically
+
+#### Master Version
+
+We store a copy of the latest version of an object's metadata using
+`object_name` as the key; this version is called the master version. The master
+versions of the objects facilitate the standard GET operation, which would
+otherwise need to scan among the list of versions of an object for its latest
+version.
+
+The following table shows the layout of all versions of `foo` in the previous
+example stored in the metadata (with dot `.` representing the null separator):
+
+| key | value |
+|-----|:-----:|
+| foo |   A   |
+| foo.v2 |   A   |
+| foo.v1 |   B   |
+
+#### Metadata Versioning Options
+
+S3 Server sends instructions to the metadata engine about whether to create a
+new version or overwrite, retrieve, or delete a specific version by sending
+values for special options in PUT, GET, or DELETE calls to metadata.
+
+Note: all operations (PUT and DELETE) that generate a new version of an object
+will return the `version_id` of the new version to the API.
+
+##### PUT
+
+- no options: original PUT operation, will update the master version
+- `versioning: true` create a new version of the object,
+  then update the master version with this version.
+- `versionId: <versionId>` update a specific version (for updating version's ACL
+  or remote updates in geo-replication);
+  - if the version identified by `versionId` happens to be the latest version,
+    the master version will be updated as well
+  - note that with `versionId` set to an empty string `''`, it will overwrite
+    the master version only (same as no options, but returns a version ID)
+
+##### DELETE
+
+-  no options: original DELETE operation, will delete the master version
+- `versionId: <versionId>` delete a specific version
+
+Version-specific deletions involve a few steps. A deletion targeting the latest
+version of an object has to:
+
+- delete the specified version identified by `versionId`
+- replace the master version with a place holder for deletion version
```

Comment:
I added some more detail; I'm not sure it's important to people who are modifying S3 though.